rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/model/ParametricFileFacts.java

```diff
@@ -108,66 +134,228 @@ public void close(ISourceLocation loc) {
+                            // calculation. In this way, a reliable diff of the
+                            // latest analyzer diagnostics and the current
+                            // builder diagnostics can be computed (by removing
+                            // the former from the latter).
+                            var build = summaryBuilder.calculateSummary(CompletableFuture.completedFuture(analysis.tree));
+                            var buildTree = build.input;
+                            var buildMessages = build.output.get().thenApply(Supplier::get);
+                            return buildTree.thenCombine(buildMessages, (tree, messages) -> {
+                                if (build.calculation.isInterrupted()) {
+                                    return Optional.<VersionedDiagnostics>empty();
+                                } else {
+                                    messages.removeAll(analysis.messages);
+                                    return Optional.of(reportDiagnostics(diagnosticsBuilder, tree, messages));
+                                }
+                            });
+                        });
+                    }
+
+                    // If the analyzer was abandoned (due to debouncing) or
+                    // interrupted, then diagnostics of the analyzer aren't
+                    // available, so retry from the start.
+                    else {
+                        calculateBuilder(version, delay);
+                    }
+                });
+            }
+        }
+
+        public ParametricSummaryBridge getSummaryAnalyzer() {
+            return summaryAnalyzer;
         }
 
-        public ParametricSummaryBridge getSummary() {
-            return summary;
+        public ParametricSummaryBridge getSummaryBuilder() {
+            return summaryBuilder;
         }
 
-        public void reportParseErrors(List<Diagnostic> msgs) {
-            parseMessages = msgs;
-            sendDiagnostics();
+        public void reportParseErrors(VersionedTree tree, List<Diagnostic> messages) {
+            reportDiagnostics(diagnosticsParser, tree, messages);
         }
 
         private void sendDiagnostics() {
             if (client == null) {
                 logger.debug("Cannot send diagnostics since the client hasn't been registered yet");
                 return;
             }
-            logger.trace("Sending diagnostics for {}. {} messages", file, typeCheckerMessages.size());
+            var messages = VersionedDiagnostics.union(diagnosticsParser.get(), diagnosticsAnalyzer.get(), diagnosticsBuilder.get());
+            logger.trace("Sending diagnostics for {}. {} messages", file, messages.size());
             client.publishDiagnostics(new PublishDiagnosticsParams(
                 file.getURI().toString(),
-                union(parseMessages, typeCheckerMessages)));
-        }
-    }
-
-    private static <T> List<T> union(List<T> a, List<T> b) {
```

Comment:
can we make the union code for 3 lists? so that we can avoid allocating lists so frequently.? especially around the parser, this happens more than we like.

Reply:
Fixed